Pitch: String processing algorithms

itingliu opened the pull request.

itingliu commented on Documentation/Evolution/StringProcessingAlgorithms.md:
````diff
@@ -0,0 +1,402 @@
+#### Trim prefix
+
+```swift
+extension Collection {
````
I find myself squinting to see the differences between these functions. I wonder if it's easier to review if we group them by extension other than algorithms, like

```swift
extension Collection where Element: Equatable {
    public func contains<S: Sequence>(_ other: S) -> Bool
    public func trimmingPrefix<Prefix: Collection>(_ prefix: Prefix) -> SubSequence
}

extension BidirectionalCollection where SubSequence == Substring {
    public func contains<R: RegexProtocol>(_ regex: R) -> Bool
    public func starts<R: RegexProtocol>(with regex: R) -> Bool
}
```

milseman commented on Documentation/Evolution/StringProcessingAlgorithms.md:
````diff
@@ -0,0 +1,402 @@
+}
+```
+
+This works, but wouldn't it be much more approachable if you can directly use the date parser within the string match function?
````
```suggestion
While this approach happens to work for this example, it is fragile and more onerous than just directly using Foundation's date parser directly within the regex.
```

I would recommend against rhetorical questions, especially "wouldn't it be better" style ones.

milseman replied:
Also, you can consider streamlining some of this as part of an extended motivation section.

milseman commented on Documentation/Evolution/StringProcessingAlgorithms.md:
````diff
@@ -0,0 +1,402 @@
+"""
+```
+
+We have already seen that parsing a date string can be tricky since it could contain localized month name (`"Feb"` as seen from above). Parsing a currency string such as `$3,020.85` with regex is not trivial either -- it can contain grouping separators, a decimal separator, and a currency symbol, all of which can be localized. 
+
+Foundation has various parsers for localized strings like these. Delegating this task to dedicated parsers alleviates the need to handle it yourself. In the second part of the pitch, we introduce the `CustomRegexComponent` protocol that conveniently lets types from outside the standard library participate in regex builders and `RegexProtocol` algorithms.
````
```suggestion
Parsing a date string can be tricky, as it could contain localized information (`"Feb"` as seen from above). Parsing a currency string such as `$3,020.85` with regex is also tricky -- it can contain grouping separators, a decimal separator, and a currency symbol, all of which can be localized. This is why Foundation provides industrial-strength parsers for localized strings like these.

 We propose a `CustomRegexComponent` protocol which allows types from outside the standard library participate in regex builders and `RegexProtocol` algorithms.
```

My suggestions are minor localized editorial changes. But, we're repeat the same information throughout, so there's more tidying up. Since this is the proposed solution section, it's a little weird to be talking about the "second part of this pitch", because this is where the second part goes.

milseman commented on Documentation/Evolution/StringProcessingAlgorithms.md:
```diff
@@ -0,0 +1,402 @@
+
+Foundation has various parsers for localized strings like these. Delegating this task to dedicated parsers alleviates the need to handle it yourself. In the second part of the pitch, we introduce the `CustomRegexComponent` protocol that conveniently lets types from outside the standard library participate in regex builders and `RegexProtocol` algorithms.
+
+If `Foundation.FloatingPointFormatStyle<Double>.Currency` conformed to `CustomRegexComponent`, you would be able to retrieve the currency from the bank statement above as a list of `Double` values this way:
```
```suggestion
`CustomRegexComponent` allows types, such as `Foundation.FloatingPointFormatStyle<Double>.Currency`, to be used directly within a regex.
```

I understand that you're not proposing the conformances for Foundation here, but I don't think the "if ... you would be able to..." adds clarity.

milseman commented on Documentation/Evolution/StringProcessingAlgorithms.md:
````diff
@@ -0,0 +1,402 @@
+}
+```
+
+You can conform your custom parser to `CustomRegexComponent`. Conformance is simple: implement the `match` function to return the upper bound of the matched substring, and the type represented by the matched range. It inherits from `RegexProtocol`, so you will be able to use it with all of the string algorithms that take a `RegexProtocol` type.
````
```suggestion
Custom parsers can conform by implementing the `match` function to return the upper bound of the matched substring, and the value constructed by the match. Conformers naturally inherit from `RegexProtocol`, so they can be used with all of the string algorithms generic over `RegexProtocol`.
```

milseman commented on Documentation/Evolution/StringProcessingAlgorithms.md:
```diff
@@ -0,0 +1,402 @@
+
+You can conform your custom parser to `CustomRegexComponent`. Conformance is simple: implement the `match` function to return the upper bound of the matched substring, and the type represented by the matched range. It inherits from `RegexProtocol`, so you will be able to use it with all of the string algorithms that take a `RegexProtocol` type.
+
+Here, we use Foundation framework's `FloatingPointFormatStyle<Double>.Currency` as an example. `FloatingPointFormatStyle<Double>.Currency` would conform to `CustomRegexComponent` by implementing the `match` function, with `Match` being a `Double`. It could also add a static function `.localizedCurrency(code:)` as a member of `RegexProtocol`, so you can refer to it as `.localizedCurrency(code:)` i ...
```
It really seems like this should be merged into the "proposed solution" sub-section, where the example's fresh in the reader's mind.

milseman commented on Documentation/Evolution/StringProcessingAlgorithms.md:
````diff
@@ -0,0 +1,402 @@
+}
+```
+
+Users could specify a pattern to match a localized currency amount such as `"$3,020.85"` simply with the following, and use it in any of the string matching algorithms introduced above.
````
```suggestion
Matching and extracting a localized currency amount, such as `"$3,020.85"`, can be done directly within a regex:
```

milseman reviewed (commented):
Just some wordsmithing and editorial adjustments. This is looking good so far, though there's opportunity to have less redundancy.

itingliu commented on Documentation/Evolution/StringProcessingAlgorithms.md:
```diff
@@ -0,0 +1,515 @@
+  
+    public func replacing<R: RegexProtocol, Replacement: Collection>(
+        _ regex: R,
+        with replacement: (RegexMatch<R.Match>) throws -> Replacement,
```
@timvermeulen @milseman I need help for documentation here. I'm not sure what this closure does.

milseman replied:
The closure receives the full match info, including captures, and returns a replacement string.

milseman commented:
@itingliu let me know when's a good time to do another pass over the prose. I don't want to conflict or interrupt any reorganizing you're doing.

rxwei commented on Documentation/Evolution/StringProcessingAlgorithms.md:
```diff
@@ -0,0 +1,545 @@
+2. Generic `Collection` equivalents of these algorithms in terms of subsequences
+3. `protocol CustomRegexComponent`, allowing libraries to vend types that can be intermixed as components of regexes
+
+This proposal is part of a larger [regex-powered string processing initiative](https://forums.swift.org/t/declarative-string-processing-overview/52459). Throughout the document, we will reference the still-in-progress [`RegexProtocol`, `Regex`](https://github.com/apple/swift-experimental-string-processing/blob/main/Documentation/Evolution/StronglyTypedCaptures.md), and [result builder DSL](https: ...
```
The "result builder DSL" is linking to the wrong pitch though, the link points to the regex literal pitch. We don't have a DSL pitch yet — I think you can just use plain text for now, or link to https://github.com/rxwei/swift-experimental-string-processing/blob/dsl-api-walkthrough/Documentation/DSLAPIWalkthough.md (which I plan to turn into a pitch very soon).

itingliu replied:
Thanks for catching that. I removed the link all together.

milseman commented on Documentation/Evolution/StringProcessingAlgorithms.md:
```diff
@@ -0,0 +1,544 @@
+
+## Motivation
+
+Below is how [Python string processing API](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str) compare to their counterparts in Swift. Many essential string processing APIs are evidently missing in the standard library. While most of those can be substituted by chaining multiple functions or using Foundation, they are basic enough to warrant a place in the standard library.
```
```suggestion
Below is a comparison of how Swift's APIs stack up with Python's. Some of the gaps can be accomplished through a series of API calls, but every gap adds a burden to developers doing frequent or complex string processing.
```

milseman replied:
Just word smithing. You can put the hyperlink back in too.

milseman commented on Documentation/Evolution/StringProcessingAlgorithms.md:
```diff
@@ -0,0 +1,544 @@
+
+Note: The comparison table omits functions to query if all characters in the string are of a specified category, such as `isalnum()` and `isalpha()`. These are achievable in Swift by passing in the corresponding character set to `allSatisfy(_:)`, so they're omitted in this table for simplicity. 
+
+### Processing domain-specific strings
```
```suggestion
### Complex string processing
```

milseman commented on Documentation/Evolution/StringProcessingAlgorithms.md:
```diff
@@ -0,0 +1,544 @@
+
+### Processing domain-specific strings
+
+Another common task regarding string processing is handling those with domain specific information. 
```
```suggestion
Even with these API additions, more complex string processing quickly becomes unwieldy. Up-coming support for authoring regexes in Swift help alleviate this somewhat, but string processing in the modern world involves dealing with localization, standards-conforming validation, and other concerns for which a dedicated parser is required.
```

milseman commented on Documentation/Evolution/StringProcessingAlgorithms.md:
````diff
@@ -0,0 +1,544 @@
+}
+```
+
+While this approach happens to work for this example, it is fragile when it comes to localized strings. 
````
```suggestion
This requires writing a simplistic pre-parser before invoking the real parser. The pre-parser will suffer from being out-of-sync and less featureful than what the real parser can do.
```

milseman commented on Documentation/Evolution/StringProcessingAlgorithms.md:
````diff
@@ -0,0 +1,544 @@
+"""
+```
+
+Parsing a currency string such as `$3,020.85` with regex is also tricky, as it can contain localized and currency symbols. This is why Foundation provides industrial-strength parsers for localized strings like these. 
````
```suggestion
Parsing a currency string such as `$3,020.85` with regex is also tricky, as it can contain localized and currency symbols in addition to accounting conventions. This is why Foundation provides industrial-strength parsers for localized strings.
```

milseman commented on Documentation/Evolution/StringProcessingAlgorithms.md:
```diff
@@ -0,0 +1,544 @@
+
+### Extend `Sequence` instead of `Collection`
+
+All of the proposed algorithms are specific to the `Collection` protocol, without support for plain `Sequence`s. Types conforming to the `Sequence` protocol are not required to support multi-pass iteration, which makes a `Sequence` conformance insufficient for most of these algorithms. In light of this, the decision was made to have the underlying shared algorithm implementations work exclusively ...
```
```suggestion
Most of the proposed algorithms are necessarily on `Collection` due to the use of indices or mutation.  `Sequence` does not support multi-pass iteration, so even `trimPrefix` would problematic on `Sequence` because it needs to look 1 `Element` ahead to know when to stop trimming.
```

milseman reviewed (approved):
Overall LGTM. I think some of the "backwards" future direction is salvageable and should be mentioned, but it doesn't need to be presented as a big deal.

After that, ship it!

itingliu commented:
> Overall LGTM. I think some of the "backwards" future direction is salvageable and should be mentioned, but it doesn't need to be presented as a big deal.
> 
> After that, ship it!

I reworked it a little bit and added it back. Thanks!

milseman commented:
@swift-ci please test
